### PR TITLE
feat: implement promotions live update

### PR DESCRIPTION
**Pull request by alexmt**
https://github.com/akuity/kargo/assets/426437/43362afb-d01a-4f8b-9b8a-c23f9f2ab83f

### CODE REVIEW

File: ui/src/features/stage/promotions.tsx [hunk 1]
```diff
@@ -1,27 +1,80 @@
...
     enabled: !!stageName
   });
 
+  useEffect(() => {
+    if (isLoading || !promotionsResponse) {
+      return;
+    }
+    const cancel = new AbortController();
+
+    const watchStages = async () => {
```
Comment (rpelczar) [inline ui/src/features/stage/promotions.tsx:36]:
```suggestion
    const watchPromotions = async () => {
```

File: ui/src/features/stage/promotions.tsx [hunk 2]
```diff
@@ -1,27 +1,80 @@
...
+
+        // Update Stages list
+        const listStagesQueryKey = listPromotions.getQueryKey({
+          project: projectName,
+          stage: stageName
+        });
+        client.setQueryData(listStagesQueryKey, { promotions });
+      }
+    };
+    watchStages();
```
Comment (rpelczar) [inline ui/src/features/stage/promotions.tsx:73]:
```suggestion
    watchPromotions();
```

**Review by rbreeze — APPROVED** [review]
Awesome LGTM!

File: ui/src/features/stage/promotions.tsx [hunk 3]
```diff
@@ -1,27 +1,80 @@
...
+            ];
+          }
+        }
+
+        // Update Stages list
+        const listStagesQueryKey = listPromotions.getQueryKey({
+          project: projectName,
+          stage: stageName
+        });
+        client.setQueryData(listStagesQueryKey, { promotions });
```
Comment (rpelczar) [inline ui/src/features/stage/promotions.tsx:70]:
```suggestion
        // Update Promotions list
        const listPromotionsQueryKey = listPromotions.getQueryKey({
          project: projectName,
          stage: stageName
        });
        client.setQueryData(listPromotionsQueryKey, { promotions });
```

**Review by jessesuen — APPROVED** [review]

File: ui/src/features/stage/promotions.tsx [hunk 4]
```diff
@@ -1,27 +1,80 @@
...
 export const Promotions = () => {
+  const client = useQueryClient();
   const { name: projectName, stageName } = useParams();
   const { data: promotionsResponse, isLoading } = useQuery({
     ...listPromotions.useQuery({ project: projectName, stage: stageName }),
     enabled: !!stageName
   });
 
+  useEffect(() => {
+    if (isLoading || !promotionsResponse) {
```
Comment (rpelczar) [inline ui/src/features/stage/promotions.tsx:31]: Is `loading` necessary here? If `promotionsResponse` is `undefined`, `isLoading` is true.

**Review by rpelczar — CHANGES_REQUESTED** [review]

**Pull request merged**